control/discovery.py

```diff
@@ -0,0 +1,958 @@
+        # port ID
+        self.port_id = bytearray(2)
+        # controller ID
+        self.controller_id = bytearray(2)
+        # admin max SQ size
+        self.asqsz = bytearray(2)
+        self.reserved1 = bytearray(22)
+        # transport service indentifier
+        self.trsvcid = bytearray(32)
+        self.reserved2 = bytearray(192)
+        # NVM subsystem qualified name
+        self.subnqn = bytearray(256)
+        # Transport address(TRADDR)
+        self.traddr = bytearray(256)
+        # Transport specific address subtype
+        self.tsas = bytearray(256)
+
+    def compose_reply(self):
+        return self.trtype + self.adrfam + self.subtype + \
+               self.treq + self.port_id + self.controller_id + \
+               self.asqsz + self.reserved1 + self.trsvcid + \
+               self.reserved2 + self.subnqn + self.traddr + self.tsas
+
+class DiscoveryService:
+    """Implements discovery controller.
+
+    Response discover request from initiator.
+
+    Instance attributes:
+        version: Discovery controller version
+        config: Basic gateway parameters
+        logger: Logger instance to track discovery controller events
+        omap_name: OMAP object name
+        ioctx: I/O context which allows OMAP access
+        discovery_addr: Discovery controller addr which allows initiator send command
+        discovery_port: Discovery controller's listening port
+    """
+
+    BDEV_PREFIX = "bdev_"
+    NAMESPACE_PREFIX = "namespace_"
+    SUBSYSTEM_PREFIX = "subsystem_"
+    HOST_PREFIX = "host_"
+    LISTENER_PREFIX = "listener_"
+
+    def __init__(self, config):
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(level=logging.DEBUG)
+
+        self.version = 1
+        self.config = config
+
+        gateway_group = self.config.get("gateway", "group")
+        self.omap_name = f"nvmeof.{gateway_group}.state" if gateway_group else "nvmeof.state"
+        self.logger.info(f"omap_name: {self.omap_name}")
+
+        ceph_pool = self.config.get("ceph", "pool")
+        ceph_conf = self.config.get("ceph", "config_file")
+        conn = rados.Rados(conffile=ceph_conf)
+        conn.connect()
+        self.ioctx = conn.open_ioctx(ceph_pool)
+
+        self.discovery_addr = self.config.get("discovery", "addr")
+        self.discovery_port = self.config.get("discovery", "port")
+        if self.discovery_addr == '' or self.discovery_port == '':
+            self.logger.error(f"discovery addr/port are empty.")
+            assert 0
+        self.logger.info(f"discovery addr: {self.discovery_addr} port: {self.discovery_port}")
+
+    def _read_all(self) -> Dict[str, str]:
+        """Reads OMAP and returns dict of all keys and values."""
+
+        with rados.ReadOpCtx() as read_op:
+            iter, _ = self.ioctx.get_omap_vals(read_op, "", "", -1)
+            self.ioctx.operate_read_op(read_op, self.omap_name)
+            omap_dict = dict(iter)
+        return omap_dict
```

Comment:
Shouldn't we perhaps start locking this object before reading/writing (here and in the gateway)? As there are cross-key dependencies (a listener maps to an NQN), unless we always enforce a specific RPC creation/deletion order, there might happen inconsistencies here. Probably nothing catastrophic, but given this is a distributed system, I'd rather be strict than performant.

@idryomov?

Reply:
> Shouldn't we perhaps start locking this object before reading/writing (here and in the gateway)? As there are cross-key dependencies (a listener maps to an NQN), unless we always enforce a specific RPC creation/deletion order, there might happen inconsistencies here.

There are cross-key dependencies but multi-key updates are atomic.  The gateway API is pretty low level so dealing with scenarios like a missing listener is a given.  But anyway, some sort of locking is exactly what the continuation of 5/23/2023 discussion that got pushed by containerization topics is about!